Add buildextend-installer command to build installer ISO

dustymabe opened the pull request.

dustymabe commented:
This is a complete WIP PR. It currently does not attempt to integrate with other outstanding PRs that may be needed to structurally build this the correct way in COSA.

- Right now you can generate an installer with `coreos-assembler buildextend-installer`.
- It expects the installer dracut module to be embedded in the ostree (currently not upstream)
- It expects to find isolinux configuration files in `/src/config/isolinux/`. For example I have `boot.msg` and `isolinux.cfg` inside of that directory on my local system where I've been hacking on this.

cgwalters commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,95 @@
+    buildmeta_commit = buildmeta['ostree-commit']
+    run_verbose(['/usr/bin/ostree', '--repo=./repo', 'checkout',
+                 '--user-mode', f"{buildmeta_commit}", 
+                 '--subpath=/usr/lib/ostree-boot', tmpostreedir])
```
You want `/usr/lib/modules` nowadays.

dustymabe replied:
cool, will look into it

dustymabe commented:
> It currently does not attempt to integrate with other outstanding PRs that may be needed to structurally build this the correct way in COSA.

this is a way of saying, let's not get too caught up on the structure just yet.

cgwalters commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,95 @@
+    run_verbose(['/usr/bin/ostree', '--repo=./repo', 'checkout',
+                 '--user-mode', f"{buildmeta_commit}", 
+                 '--subpath=/usr/lib/ostree-boot', tmpostreedir])
+    run_verbose('/usr/bin/mv {} {}'.format(
```
While this is OK here passing filenames into the shell is a general source of problems (and for e.g. web apps, security vulnerabilities).  

Related to the above, what you want is the first non-empty dir in `/usr/lib/modules/` and then it's just `vmlinuz` and `initramfs.img`.  See the libostree source related to this.

cgwalters commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,95 @@
+                shell=True)
+
+    # Grab all the contents from the isolinux dir from the configs
+    run_verbose(f"/usr/bin/cp ./src/config/isolinux/* {tmpisoroot}", shell=True)
```
Maybe cleaner as:
`run_verbose(["rsync", "-a", "src/config/isolinux/", f"{tmpisoroot}/"])`

dustymabe commented:
at least those few comments should be addressed. I've got to take a look at the other PRs and see where this all fits in. Also I just created the github.com/coreos/coreos/installer repo where I'll be posting up the installer code for this.

dustymabe commented:
> > I've got to take a look at the other PRs and see where this all fits in
> 
> Luckily given we didn't explicitly coordinate, I believe this is actually nicely complementary to my work in
> #305

Nice. Sorry I haven't looked at it yet. I'm going to get the bits up at github.com/coreos/coreos-installer today and then have a look at the outstanding PRs you have open here.

> 
> This is about a PXE path (right)? Which will end up `dd`-ing an image to disk. I assume you've been testing with the "base image" which today is cloud oriented. My PR is about having a base disk image designed for metal.

Yeah I've been just using the disk image that we've been building (converting it to raw first).

cgwalters commented:
So this ISO will boot into the initramfs; in order to do an install one will need to change the kernel command line to point to the target disk image, right?

(Soon we're going to need to cross the "documentation for FCOS" issue)

cgwalters commented:
Code seems sane to me!

dustymabe commented:
OK I've done a little knocking down some of these roadblocks:

>  Right now you can generate an installer with `coreos-assembler buildextend-installer`.

This is this PR. You'll need to pull down this PR to use it.

>  It expects the installer dracut module to be embedded in the ostree (currently not upstream)

I've upstreamed my work to https://github.com/coreos/coreos-installer/tree/master

>  It expects to find isolinux configuration files in `/src/config/isolinux/`. For example I have `boot.msg` and `isolinux.cfg` inside of that directory on my local system where I've been hacking on this.

Opened a PR here for this: https://github.com/coreos/fedora-coreos-config/pull/47. As part of that PR I also have a section that includes the dracut files directly into the OSTree for fast development (see https://github.com/coreos/fedora-coreos-config/pull/47/commits/a8577b86cd90bcf060bd5b181aeb902ea18517bb).

You should be able to:

```
$ cd $COREOS_ASSEMBLER_CONFIG_GIT
$ git fetch origin pull/47/head:PR47; git checkout PR47
$ git clone https://github.com/coreos/coreos-installer
$ cd -
$ coreos-assembler build
$ coreos-assembler buildextend-installer
```

You'll then need to copy the generated ISO somewhere and boot it. You can provide command line args: `coreos.inst.image_url` `coreos.inst.install_dev` `coreos.inst.ignition_url`.

This is still very rough around the edges so I only suggest trying this if you like pain and can't wait :) 

Also noteworthy is that ignition fails creating the coreos user when trying to actually boot the generated FCOS qcow with the installer dracut module included. Haven't had a chance to debug that one yet.

ashcrow commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,96 @@
+        os.chmod(dst, mode)
+
+    # Generate the ISO image
+    run_verbose(['/usr/bin/genisoimage', '-b', 'isolinux.bin', '-c', 'boot.cat',
```
I'm guessing the `genisoimage` is pulled in as a dependency of another package. It may be worth adding it to the `deps.txt`.

dustymabe replied:
👍 - fixed

ashcrow commented:
/cc @mrguitar

dustymabe commented:
fixed code review comments, rebased to latest upstream master, and also added an exception to `cosa compress` to not compress ISO images.

I'm going to lift WIP for now. The other work is still in PRs elsewhere but I think we can go ahead and merge this without bothering anything else.

cgwalters commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,96 @@
+                 '-no-emul-boot', '-boot-load-size', '4', '-boot-info-table',
+                 '-rock', '-J', '--verbose', '-o', tmpisofile, tmpisoroot])
+    checksum = sha256sum_file(tmpisofile)
+    buildmeta['images']['iso'] = {
```
Eventually we'll want to land a `live-iso` or so right?  How about `installer-iso`?

dustymabe replied:
I think we're still working out those details as we may just decide to ship one ISO that can do both. Still some details to work out there. Let's cross that in a followup if you don't mind.

jlebon commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,96 @@
+builddir = os.path.join('builds', args.build)
+buildmeta_path = os.path.join(builddir, 'meta.json')
+with open(buildmeta_path) as f:
+    buildmeta = json.load(f)
```
Might be worth checking the metadata if an ISO was already generated so the command is idempotent.

dustymabe replied:
done!

dustymabe commented:
> How about a commit message like:
> 
> ```
> buildextend-installer: New command
> 
> This will allow people to use the new installer
> https://github.com/coreos/coreos-installer/
> to e.g. do PXE installs to bare metal.
> ```

done. though I `s/PXE/ISO/` for now since I haven't specifically addressed that use case yet

dustymabe commented:
added fixups from CR and rebased on latest master

jlebon commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,106 @@
+parser = argparse.ArgumentParser()
+parser.add_argument("--build", help="Build ID")
+parser.add_argument("--force", action='store_true', default=False,
+                    help="Overwrite Previously Generated Installer")
```
Minor/optional: help text isn't usually Title Case.

jlebon commented on src/cmd-buildextend-installer:
```diff
@@ -0,0 +1,106 @@
+# Don't run if it's already been done, unless forced
+if 'iso' in buildmeta['images'] and not args.force:
+    print(f"Installer has already been built for {args.build}. Skipping") 
+    print("You can force a rebuild with --force")
```
Minor/optional: these sentences are missing their final periods.

dustymabe replied:
fixed

dustymabe commented:
fixed CR comments from @jlebon

jlebon commented:
LGTM.
I just tweaked the PR title to include the new command name to make it easier to find in the future.

dustymabe commented:
> I just tweaked the PR title to include the new command name to make it easier to find in the future.

👍  - anyone opposed to merging?

jlebon commented:
Will merge before end of day today if there are no other objections.
Awesome work on this BTW! :+1:

mrguitar commented:
Yeah nice work on this. Only issue I see is I don't think this will work w/ UEFI support, but that should be pretty simple to add later.
